File: framework/languages/java/e2e/RegistryMessaging/Server/Provider/src/main/java/systems/zlink/e2e/registrymessaging/provider/Configuration/ServerOptions.java
package systems.zlink.e2e.registrymessaging.provider.Configuration;

import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties("e2e")
public record ServerOptions(
    String providerRid,
    String providerInstance,
    String apiWeight,
    String apiEndpoint,
    String routeEndpoint,
    String routePeers,
    String routePeerRids,
    String workflowEndpoint,
    int httpPort,
    String redisLocationEndpoint,
    String locationKeyPrefix,
    String logDir) {
    public ServerOptions {
        required(providerRid, "provider-rid");
        required(providerInstance, "provider-instance");
        required(redisLocationEndpoint, "redis-location-endpoint");
        required(locationKeyPrefix, "location-key-prefix");
        required(logDir, "log-dir");
        apiWeight = optional(apiWeight);
        apiEndpoint = optional(apiEndpoint);
        routeEndpoint = optional(routeEndpoint);
        routePeers = optional(routePeers);
        routePeerRids = optional(routePeerRids);
        workflowEndpoint = optional(workflowEndpoint);
    }

    private static String optional(String value) { return value == null ? "" : value; }
    private static void required(String value, String name) {
        if (value == null || value.isBlank()) throw new IllegalArgumentException("e2e." + name + " is required");
    }
}

File: framework/languages/java/e2e-kotlin/RegistryMessaging/Server/Provider/src/main/kotlin/systems/zlink/e2e/kotlin/registrymessaging/provider/Configuration/ServerOptions.kt
package systems.zlink.e2e.kotlin.registrymessaging.provider.Configuration

data class ServerOptions(
    val rid: String,
    val instanceId: String,
    val httpUrl: String,
    val logDir: String,
    val evidenceFile: String?,
    val redisLocationEndpoint: String,
    val locationKeyPrefix: String,
    val channelEndpoint: String?,
    val routeEndpoint: String?,
    val routePeerRids: List<String>,
    val weight: Int,
    val routePeers: List<String>,
) {
    companion object {
        fun parse(args: Array<String>): ServerOptions {
            val values = linkedMapOf<String, MutableList<String>>()
            var index = 0
            while (index < args.size) {
                val key = args[index]
                if (!key.startsWith("--")) {
                    index++
                    continue
                }
                require(index + 1 < args.size) { "Missing value for $key." }
                values.getOrPut(key.removePrefix("--")) { mutableListOf() }.add(args[index + 1])
                index += 2
            }
            val rid = values.last("rid", "node")
            return ServerOptions(
                rid = rid,
                instanceId = values.last("instance-id", rid),
                httpUrl = values.last("http-url", "http://127.0.0.1:0"),
                logDir = values.last("log-dir", "logs"),
                evidenceFile = values.lastOrNull("evidence-file"),
                redisLocationEndpoint = values.lastRequired("redis-location-endpoint"),
                locationKeyPrefix = values.lastRequired("location-key-prefix"),
                channelEndpoint = values.lastOrNull("channel-endpoint"),
                routeEndpoint = values.lastOrNull("route-endpoint"),
                routePeerRids = values["route-peer-rid"]?.toList().orEmpty(),
                weight = values.last("weight", "100").toInt(),
                routePeers = values["route-peer"]?.toList().orEmpty(),
            )
        }
    }
}

private fun Map<String, List<String>>.last(key: String, defaultValue: String): String =
    this[key]?.lastOrNull()?.takeIf { it.isNotBlank() } ?: defaultValue

private fun Map<String, List<String>>.lastOrNull(key: String): String? =
    this[key]?.lastOrNull()?.takeIf { it.isNotBlank() }

private fun Map<String, List<String>>.lastRequired(key: String): String =
    lastOrNull(key) ?: throw IllegalArgumentException("--$key is required.")
